### ISSUE TITLE
Create local development environment setup

### ISSUE THREAD

**Issue by aidandunlop** [sky-uk/kfp-operator#380]
**Is your feature request related to a problem? Please describe.**
The `integration-test-up`/`integration-test` makefile targets are currently the only way to run the operator locally, but they only test the Argo workflows. It would be great to have a makefile target that spins up a local kubernetes cluster using `minikube`, with all the necessary `CustomResourceDefinition`s installed, along with the eventing system and the operator itself. A local provider could be set such that it does not need to connect to any external APIs.

**Describe the solution you'd like**
- single makefile target that can setup a minikube cluster with the operator installed and all the necessary dependencies to be able to apply `pipelines.kubeflow` custom resources.
- single makefile target to cleanly remove the above minikube cluster
- need to consider running the compiler on arm macs, as previously we have encountered issues with this
- utility to install all asdf plugins required in one make command
- We can create mock services for external dependencies:
  - create a mock for VAI (including mock of the methods/interfaces/grpc Services we use from VAI: https://cloud.google.com/go/docs/reference/cloud.google.com/go/aiplatform/latest/apiv1)
    - ScheduleServiceClient, PipelineJob etc
  - use [fake-gcs-server](https://github.com/fsouza/fake-gcs-server)
  - use mock pub sub
- we should be able to run Kubeflow Pipelines locally

**Describe alternatives you've considered**
- improve the integration tests to more scenarios

**Comment by aidandunlop** [comment]
Some work has begun on this issue in the `local-dev-env` [branch](https://github.com/sky-uk/kfp-operator/tree/local-dev-env)

**Comment by TobyPinfold** [comment]
Decided to pause work on this issue due to M1 compatibility issues. Plan is to revisit once TFX supports TF >=2.16.

Current investigation steps for future work:

## Prerequisites:
 - Minikube
 - Helm

## Setting up minikube for KFP
-  `minikube addons enable registry` Enable the registry addon to easily push images for use within minikube
- `minikube start --driver=docker` 
-  `docker run -d --rm -it --network=host --name minikube-registry-proxy alpine ash -c "apk add socat && socat TCP-LISTEN:5000,reuseaddr,fork TCP:$(minikube ip):5000"` Set up proxy to minikubes registry exposed on localhost.

## Install argo
-  `helm repo add argo https://argoproj.github.io/argo-helm` Add argo repo to helm
-  `helm install argo-workflows argo/argo-workflows -n argo --create-namespace` Install argo-workflows
-  `helm install argo-events argo/argo-events -n argo-events --create-namespace` Install argo-events

## Install cert-manager and configure certificates
- `kubectl apply -f https://github.com/cert-manager/cert-manager/releases/download/v1.9.1/cert-manager.crds.yaml` Install cert manager.
- `openssl req -newkey rsa:2048 -nodes -keyout webhook-server-key.pem -x509 -days 365 -out webhook-server-cert.pem -subj "/CN=webhook-server.kubeflow.svc"` Create certificates.
- `create secret tls webhook-server-cert --cert=webhook-server-cert.pem --key=webhook-server-key.pem --namespace kfp-operator-system` Add certificates to cluster.

## Deploy KFP operator
- `CONTAINER_REPOSITORIES=localhost:5000 make docker-push docker-push-argo docker-push-triggers` Push KFP images to minikube registry
- ensure helm values has `containerRegistry` set to `localhost:5000`
- `make helm-install-operator` 

## Teardown
- `minikube stop && minikube delete`
- `docker rm minikube-registry-proxy`

**Comment by grahamia** [comment]
KFP Operator itself installs correctly now into minikube.

Next step is probably to get the provider CLI at least working and see if possible to run actual VAI provider locally against own custom subscription etc. Will need to change log router to be custom but wanted to do that based on platform anyway.

Provider service stub can be created once Step 4 is complete.

**Comment by chr12c** [comment]
helm chart for the validating webhook configuration needs to be correctly configured:
- supplied `values.yaml` cannot have `manager.webhookCertificates.provider = cert-manager`. Either set this to `custom` or something else.
- at the very least, the validating webhook configuration need to have the `clientConfig.caBundle` to have a base64 encoded string of the generated cert. In this case it's `local/kfp-operator-webhook.crt`
  - if the cert is generated after the helm chart creates the resources then the caBundle would have to be patched in

**Comment by chr12c** [comment]
Regarding the branch `380-local-dev-env`

helm chart for the validating webhook configuration needs to be correctly configured:
- supplied `values.yaml` cannot have `manager.webhookCertificates.provider = cert-manager`. Either set this to `custom` or something else.
- at the very least, the validating webhook configuration need to have the `clientConfig.caBundle` to have a base64 encoded string of the generated cert. In this case it's `local/kfp-operator-webhook.crt`
  - if the cert is generated after the helm chart creates the resources then the caBundle would have to be patched in
- the `quickstart` and `kfp-operator-argo-kfp-compiler` images need to be loaded into minikube

After having addressed these issues manually, I get the following after applying an mlrc:
```
8m54s       Warning   WorkflowFailed          workflow/pipeline-quickstart-gjsjh                                   child 'pipeline-quickstart-gjsjh-1181987344' failed
8m54s       Warning   WorkflowNodeFailed      workflow/pipeline-quickstart-gjsjh                                   Failed node pipeline-quickstart-gjsjh[3]: child 'pipeline-quickstart-gjsjh-1181987344' failed
8m54s       Warning   WorkflowNodeFailed      workflow/pipeline-quickstart-gjsjh                                   Failed node pipeline-quickstart-gjsjh: child 'pipeline-quickstart-gjsjh-1181987344' failed
8m54s       Normal    WorkflowNodeRunning     workflow/pipeline-quickstart-gjsjh                                   Running node pipeline-quickstart-gjsjh[3].compile: Error (exit code 134)
8m54s       Warning   WorkflowNodeFailed      workflow/pipeline-quickstart-gjsjh                                   Failed node pipeline-quickstart-gjsjh[3].compile: Error (exit code 134)
3m27s       Warning   FailedScheduling        pod/eventbus-default-stan-2                                          0/1 nodes are available: 1 Insufficient cpu. preemption: 0/1 nodes are available: 1 No preemption victims found for incoming pod.
```

Need to see if starting up minikube with more resources will help it.

**Comment by peter-hazell** [comment]
**Note about installing cert-manager and configuring certs**

The approach in @TobyPinfold's comment doesn't work as K8s doesn't allow certs using `CN` and requires using SANs, so we moved to this approach:

```
	kubectl apply -f https://github.com/cert-manager/cert-manager/releases/download/v1.9.1/cert-manager.crds.yaml
	openssl req -new -newkey rsa:2048 -days 365 -keyout ./local/kfp-operator-webhook.key -out ./local/kfp-operator-webhook.csr \
	  -subj "/C=US/ST=California/L=San Francisco/O=My Organization/OU=My Unit/CN=kfp-operator-webhook-service.kfp-operator-system.svc" \
	  -extensions san -config <(echo "[req]"; echo "distinguished_name=req_distinguished_name"; echo "x509_extensions = san"; \
	  echo "[req_distinguished_name]"; echo "C=US"; echo "ST=California"; echo "L=San Francisco"; echo "O=My Organization"; \
	  echo "OU=My Unit"; echo "CN=kfp-operator-webhook-service.kfp-operator-system.svc"; \
	  echo "[san]"; echo "subjectAltName=DNS:kfp-operator-webhook-service.kfp-operator-system.svc,DNS:kfp-operator-webhook-service") -nodes
	openssl x509 -req -in ./local/kfp-operator-webhook.csr -signkey ./local/kfp-operator-webhook.key -out ./local/kfp-operator-webhook.crt \
	  -extensions v3_req -extfile <(echo "[v3_req]"; echo "subjectAltName=DNS:kfp-operator-webhook-service.kfp-operator-system.svc,DNS:kfp-operator-webhook-service")
	kubectl create secret tls webhook-server-cert --cert=./local/kfp-operator-webhook.crt --key=./local/kfp-operator-webhook.key --namespace kfp-operator-system
```

**Comment by peter-hazell** [comment]
Closing this issue - improvements have been made in this area and running the operator locally is possible with documentation in place as well. We should continually improve this as we work on other issues rather than having an 'always open' issue like this one, as the end goal isn't clear.
